### DISCUSSION TITLE
Guidance Request - Using Microsoft.Identity.Web.GraphServiceClient With Certificate Avoiding Appsettings.json

### DISCUSSION THREAD

**Discussion by Sn3akyP3t3** [microsoftgraph/msgraph-sdk-dotnet#2367] [Q&A]
I'm looking for advisement on how to use the `Microsoft.Identity.Web.GraphServiceClient` library, but not using appsettings.json as seen in the sample code exhibited in the tutorials and documentation.

I'm trying to follow the [advisement to use `Microsoft.Identity.Web.GraphServiceClient`](https://learn.microsoft.com/en-us/entra/identity-platform/scenario-daemon-app-configuration?tabs=idweb#instantiate-the-confidential-client-application-with-a-client-secret) instead of the tried and true `MSAL.NET` for authentication, but the sample code and all other examples seen rely on the existence of appsettings.json which I cannot adopt.  In my current application the Client and Tenant ID are set as environment variables and expected to be passed into the application that way.  The certificate is loaded from the Windows certificate store and expected to be passed, but also on Linux from file and expected passed as an object as well.

I did find another [example of what is recommended for a daemon C# application](https://learn.microsoft.com/en-us/graph/sdks/choose-authentication-providers?tabs=csharp#using-a-client-certificate), but it seems to rely on `Azure.Identity` which in turn I believe relies on MSAL.NET and not the advised `Microsoft.Identity.Web.GraphServiceClient` library.

Alternatively, I do like that another option exists ([Access Microsoft Graph from a secured app as the app](https://learn.microsoft.com/en-us/entra/identity-platform/multi-service-web-app-access-microsoft-graph-as-app?tabs=azure-powershell%2Cprogramming-language-csharp)), but I think its too soon for me to adopt this for my existing application since it will require quite a bit of refactoring and coordination with system admins to enable.

Lastly, to Microsoft.  I don't understand why you would [advise to use certificates over secrets](https://learn.microsoft.com/en-us/graph/applications-how-to-add-certificate?tabs=http#prerequisites), but then not provide an abundance of examples and sample code exhibiting use of certificates in documentation and sample code.

**Comment by pschaeflein** [comment]
The `AddMicrosoftGraph()` method has an overload that accepts `IConfiguration`.  You can set configuration in many ways other than appSettings.json, including environment: https://learn.microsoft.com/en-us/dotnet/core/extensions/configuration
